tdrs-backend/tdpservice/parsers/base_parser.py

```diff
@@ -0,0 +1,185 @@
+"""Base parser logic associated to all parser classes."""
+
+from abc import ABC, abstractmethod
+from django.conf import settings
+from django.db.utils import DatabaseError
+from elasticsearch.exceptions import ElasticsearchException
+import itertools
+import logging
+from tdpservice.parsers import util
+from tdpservice.parsers.decoders import DecoderFactory
+from tdpservice.parsers.models import ParserErrorCategoryChoices, ParserError
+from tdpservice.parsers.schema_manager import SchemaManager
+from tdpservice.parsers.util import SortedRecords, log_parser_exception
+
+logger = logging.getLogger(__name__)
+
+class BaseParser(ABC):
+    """Abstract base class for all parsers."""
+
+    def __init__(self, datafile, dfs, section):
+        super().__init__()
+        self.datafile = datafile
+        self.dfs = dfs
+        self.section = section
+        self.program_type = None
+
+        # Initialized decoder.
+        self.decoder = DecoderFactory.get_instance(datafile.file)
+
+        self.current_row = None
+        self.current_row_num = 0
+        self.errors = dict()
+
+        # Specifying unsaved_records here may or may not work for FRA files. If not, we can move it down the
+        # inheritance hierarchy.
+        self.unsaved_records = SortedRecords(section)
+        self.unsaved_parser_errors = dict()
+        self.num_errors = 0
+
+    @abstractmethod
+    def parse_and_validate(self) -> dict:
+        """To be overriden in child class."""
+        # Should have the same return as parse.py::parse_datafile
+        pass
+
+    def _init_schema_manager(self, program_type):
+        """Initialize the schema manager with the given program type."""
+        # program_type is manipulated by some parsers. E.g. the tdr_parser has to prefix a tribal TANF's program type
+        # with "Tribal" since it's base program type would just be TANF.
+        self.program_type = program_type
+        self.schema_manager = SchemaManager(self.datafile, self.program_type, self.section)
+
+    def bulk_create_records(self, header_count, flush=False):
+        """Bulk create unsaved_records."""
+        batch_size = settings.BULK_CREATE_BATCH_SIZE
+        if (self.current_row_num % batch_size == 0 and header_count > 0) or flush:
+            logger.debug("Bulk creating records.")
+            num_db_records_created = 0
+            num_expected_db_records = 0
+            num_elastic_records_created = 0
+            for document, records in self.unsaved_records.get_bulk_create_struct().items():
+                try:
+                    num_expected_db_records += len(records)
+                    created_objs = document.Django.model.objects.bulk_create(records)
+                    num_db_records_created += len(created_objs)
+                    num_elastic_records_created += document.update(created_objs)[0]
+                except ElasticsearchException as e:
```

Comment:
since the only action different is the log message and error is represented in e, I lean to one generic error instead